File: app/src/main/java/ru/nikitazhelonkin/coinbalance/data/entity/Exchange.java
package ru.nikitazhelonkin.coinbalance.data.entity;


import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.Index;
import android.arch.persistence.room.PrimaryKey;
import android.arch.persistence.room.TypeConverters;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import ru.nikitazhelonkin.coinbalance.data.db.ExchangeServiceConverter;

@Entity(indices =
@Index(name = "exchange_index", value = {"service", "api_key"}, unique = true))
public class Exchange implements ListItem {

    public static final int STATUS_NONE = 0;
    public static final int STATUS_ERROR = -1;
    public static final int STATUS_OK = 1;

    @PrimaryKey(autoGenerate = true)
    private final int mId;
    @TypeConverters(ExchangeServiceConverter.class)
    @ColumnInfo(name = "service")
    private final ExchangeService mService;
    @ColumnInfo(name = "api_key")
    private final String mApiKey;
    @ColumnInfo(name = "api_secret")
    private final String mApiSecret;
    @ColumnInfo(name = "title")
    private String mTitle;
    @ColumnInfo(name = "position")
    private int mPosition;
    @ColumnInfo(name = "created_at")
    private final long mCreatedAt;
    @ColumnInfo(name = "status")
    private int mStatus;
    @ColumnInfo(name = "error_message")
    private String mErrorMessage;

    @Ignore
    public Exchange(@NonNull ExchangeService service,
                    @NonNull String apiKey,
                    @NonNull String apiSecret,
                    @Nullable String title) {
        this(0, service, apiKey, apiSecret, title, -1, System.currentTimeMillis(), STATUS_NONE);
    }

    public Exchange(int id,
                    @NonNull ExchangeService service,
                    @NonNull String apiKey,
                    @NonNull String apiSecret,
                    @Nullable String title,
                    int position,
                    long createdAt,
                    int status) {
        mId = id;
        mService = service;
        mApiKey = apiKey;
        mApiSecret = apiSecret;
        mTitle = title;
        mPosition = position;
        mCreatedAt = createdAt;
        mStatus = status;
    }

    public int getId() {
        return mId;
    }

    public ExchangeService getService() {
        return mService;
    }

    public String getApiKey() {
        return mApiKey;
    }

    public String getApiSecret() {
        return mApiSecret;
    }

    public String getTitle() {
        return mTitle;
    }

    public void setTitle(String title) {
        mTitle = title;
    }

    public int getPosition() {
        return mPosition;
    }

    public void setPosition(int position) {
        mPosition = position;
    }

    public long getCreatedAt() {
        return mCreatedAt;
    }

    public int getStatus() {
        return mStatus;
    }

    public void setStatus(int status) {
        mStatus = status;
    }

    public void setErrorMessage(String errorMessage) {
        mErrorMessage = errorMessage;
    }

    public String getErrorMessage() {
        return mErrorMessage;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Exchange exchange = (Exchange) o;

        return mId == exchange.mId;
    }

    @Override
    public int hashCode() {
        return mId;
    }

    @Override
    public String toString() {
        return "Exchange{" +
                "mId=" + mId +
                ", mService=" + mService +
                ", mTitle='" + mTitle + '\'' +
                '}';
    }

}

File: app/src/main/java/ru/nikitazhelonkin/coinbalance/data/entity/Exchange.kt
package ru.nikitazhelonkin.coinbalance.data.entity


import android.arch.persistence.room.*
import ru.nikitazhelonkin.coinbalance.data.db.ExchangeServiceConverter

@Entity(indices = [(Index(name = "exchange_index", value = arrayOf("service", "api_key"), unique = true))])
class Exchange(@field:PrimaryKey(autoGenerate = true) @field:ColumnInfo(name = "mId") override val id: Int,
                    @field:TypeConverters(ExchangeServiceConverter::class)
                    @field:ColumnInfo(name = "service") val service: ExchangeService?,
                    @field:ColumnInfo(name = "api_key") val apiKey: String?,
                    @field:ColumnInfo(name = "api_secret") val apiSecret: String?,
                    @field:ColumnInfo(name = "title") var title: String?,
                    @field:ColumnInfo(name = "position") override var position: Int = -1,
                    @field:ColumnInfo(name = "created_at") override val createdAt: Long = System.currentTimeMillis(),
                    @field:ColumnInfo(name = "status") var status: Int = STATUS_NONE,
                    @field:ColumnInfo(name = "error_message") var errorMessage: String? = null) : ListItem {

    @Ignore
    constructor(service: ExchangeService, apiKey: String, apiSecret: String, title: String?) :
            this(id = 0, service = service, apiKey = apiKey, apiSecret = apiSecret, title = title)


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val exchange = other as Exchange?

        return id == exchange!!.id
    }

    override fun hashCode(): Int {
        return id
    }

    override fun toString(): String {
        return "Exchange(mId=$id, service=$service, title=$title)"
    }

    companion object {
        const val STATUS_NONE = 0
        const val STATUS_ERROR = -1
        const val STATUS_OK = 1
    }

}
